cmd/tempo-vulture/main.go

```diff
@@ -45,95 +42,148 @@ func init() {
 
-				// query loki for trace ids
-				lines, err := queryLoki(lokiBaseURL, lokiQuery, duration, lokiUser, lokiPass)
-				if err != nil {
-					glog.Error("error querying Loki ", err)
-					metricErrorTotal.Inc()
-					continue
-				}
-				ids := extractTraceIDs(lines)
-
-				// query tempo for trace ids
-				metrics, err := queryTempoAndAnalyze(tempoBaseURL, tempoBackoffDuration, ids)
-				if err != nil {
-					glog.Error("error querying Tempo ", err)
-					metricErrorTotal.Inc()
-					continue
-				}
+			c, err := newJaegerGRPCClient(tempoPushURL)
+			if err != nil {
+				glog.Error("error creating grpc client", err)
+				metricErrorTotal.Inc()
+				continue
+			}
 
-				metricTracesInspected.WithLabelValues(strconv.Itoa(int(duration.Seconds()))).Add(float64(metrics.requested))
-				metricTracesErrors.WithLabelValues("notfound", strconv.Itoa(int(duration.Seconds()))).Add(float64(metrics.notfound))
-				metricTracesErrors.WithLabelValues("missingspans", strconv.Itoa(int(duration.Seconds()))).Add(float64(metrics.missingSpans))
+			batch := makeThriftBatch()
+			err = c.EmitBatch(context.Background(), batch)
+			if err != nil {
+				glog.Error("error pushing trace to Tempo ", err)
+				metricErrorTotal.Inc()
+				continue
 			}
 		}
 	}()
 
+	// Read
+	go func() {
+		for {
+			<-ticker.C
+
+			if iteration == 1 {
+				continue
+			}
+
+			// pick past iteration and re-generate trace
+			rand.Seed(startTime / randInt(1, iteration-1))
+			batch := makeThriftBatch()
+			hexID := fmt.Sprintf("%016x%016x", batch.Spans[0].TraceIdHigh, batch.Spans[0].TraceIdLow)
+
+			// query the trace
+			metrics, err := queryTempoAndAnalyze(tempoQueryURL, hexID)
+			if err != nil {
+				glog.Error("error querying Tempo ", err)
+				metricErrorTotal.Inc()
+				continue
+			}
+
+			metricTracesInspected.Add(float64(metrics.requested))
+			metricTracesErrors.WithLabelValues("notfound").Add(float64(metrics.notfound))
+			metricTracesErrors.WithLabelValues("missingspans").Add(float64(metrics.missingSpans))
+		}
+	}()
+
 	http.Handle(prometheusPath, promhttp.Handler())
 	log.Fatal(http.ListenAndServe(prometheusListenAddress, nil))
 }
 
-func queryTempoAndAnalyze(baseURL string, backoff time.Duration, traceIDs []string) (*traceMetrics, error) {
-	tm := &traceMetrics{
-		requested: len(traceIDs),
+func newJaegerGRPCClient(endpoint string) (*jaeger_grpc.Reporter, error) {
+	// remove scheme and port
+	u, err := url.Parse(endpoint)
+	if err != nil {
+		return nil, err
 	}
+	host, _, err := net.SplitHostPort(u.Host)
+	if err != nil {
+		return nil, err
+	}
+	// new jaeger grpc exporter
+	conn, err := grpc.Dial(host+":14250", grpc.WithInsecure())
+	if err != nil {
+		return nil, err
+	}
+	logger, err := zap.NewDevelopment()
+	if err != nil {
+		return nil, err
+	}
+	return jaeger_grpc.NewReporter(conn, nil, logger), err
+}
 
-	for _, id := range traceIDs {
-		time.Sleep(backoff)
+func makeThriftBatch() *thrift.Batch {
+	var spans []*thrift.Span
```

Comment:
oh man, let's make some beefy traces.  100s, 1000s, 10,000s of spans split across a bunch of batches.
randomize everything names, tags, logs.

Reply:
Makes sense, I'll work on this :)

Reply:
Added the code required for creating **beefy traces**. As you can see, I've some hardcoded ranges. Probably it's interesting to let the user configure some of those (e.g. number of batches).